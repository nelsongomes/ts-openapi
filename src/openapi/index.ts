import Joi, { Schema } from "joi";

export { Joi, Schema };

export { OpenApi } from "./openapi";
export { WebRequestSchema, ParameterIn, Parameters } from "./openapi.types";

export { textPlain } from "./openapi-helpers";
