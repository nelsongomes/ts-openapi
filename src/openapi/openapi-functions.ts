import Joi, { ObjectSchema } from "joi";
import joiToSwagger from "../joi-conversion";
import {
  Body,
  IntegerFormats,
  NumberFormats,
  SchemaTypeArray,
  SchemaTypeBoolean,
  SchemaTypeInteger,
  SchemaTypeNumber,
  SchemaTypeObject,
  SchemaTypeString,
  StringFormats
} from "./openapi.types";

export function limitations(parameter: any): string {
  const limitationArray = [];

  if (parameter.type === "string" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minLength;
    delete parameter.maxLength;
  }

  // string
  if (parameter.minLength) {
    limitationArray.push(`min:${parameter.minLength} chars`);
  }

  if (parameter.maxLength) {
    limitationArray.push(`max:${parameter.maxLength} chars`);
  }

  // integer / float
  if (parameter.type === "integer" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minimum;
    delete parameter.maximum;
  }

  if (typeof parameter.minimum === "number") {
    limitationArray.push(`min:${parameter.minimum}`);
  }

  if (typeof parameter.maximum === "number") {
    limitationArray.push(`max:${parameter.maximum}`);
  }

  // array
  if (parameter.minItems) {
    limitationArray.push(`minItems:${parameter.minItems}`);
  }

  if (parameter.maxItems) {
    limitationArray.push(`maxItems:${parameter.maxItems}`);
  }

  if (parameter.format) {
    switch (parameter.format) {
      case StringFormats.DateTime:
        if (parameter.meta && parameter.meta.format === "date") {
          limitationArray.push(`date:yyyy-mm-dd`);
        } else {
          limitationArray.push(`ISO8601 date-time format`);
        }
        break;
      case StringFormats.Byte:
        limitationArray.push("base64 encoded string");
        break;
      case StringFormats.Binary:
        limitationArray.push("binary string");
        break;
    }
  }

  return limitationArray.length > 0 ? ` (${limitationArray.join(", ")})` : "";
}

export function bodyParams(schema: ObjectSchema): Body {
  const internalSchema = Joi.object().keys({ object: schema.required() });
  const query = joiToSwagger(internalSchema, {});

  const key = Object.keys(query.swagger.properties)[0];
  const parameter = query.swagger.properties[key];

  return {
    description:
      query.swagger.properties[key].description ||
      "Body does not have a description.",
    content: {
      "application/json": {
        schema: objectSchema(parameter),
        ...(parameter.example && { example: parameter.example })
      }
    }
  };
}

function objectSchema(parameter: any): SchemaTypeObject {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeObject = {
    description,
    ...(typeof parameter.default === "object" && {
      default: parameter.default
    }),
    ...(parameter.minItems && { minItems: parameter.minItems }),
    ...(parameter.maxItems && { maxItems: parameter.maxItems }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "object",
    properties: {}
  };

  for (const propertyKey of Object.keys(parameter.properties)) {
    const property = parameter.properties[propertyKey];

    switch (property.type) {
      case "string":
        const {
          default: ignoreA1,
          nullable: ignoreA2,
          minLength,
          maxLength,
          ...restOfStringSchema
        } = stringSchema(property);

        output.properties[propertyKey] = restOfStringSchema;
        break;
      case "number":
        const {
          default: ignoreB1,
          nullable: ignoreB2,
          minimum: ignoreB3,
          maximum: ignoreB4,
          ...restNumberOfSchema
        } = numberSchema(property);

        output.properties[propertyKey] = restNumberOfSchema;
        break;
      case "boolean":
        const {
          default: ignoreE1,
          nullable: ignorE2,
          ...restBooleanOfSchema
        } = booleanSchema(property);

        output.properties[propertyKey] = restBooleanOfSchema;
        break;
      case "integer":
        const {
          default: ignoreC1,
          nullable: ignoreC2,
          minimum: ignoreC3,
          maximum: ignoreC4,
          ...restIntegerOfSchema
        } = integerSchema(property);

        output.properties[propertyKey] = restIntegerOfSchema;
        break;
      case "object":
        const {
          default: ignoreD1,
          nullable: ignoreD2,
          ...restOfObjectSchema
        } = objectSchema(property);

        output.properties[propertyKey] = restOfObjectSchema;
        break;
      case "array":
        const {
          default: ignoreF1,
          nullable: ignoreF2,
          ...restOfArraySchema
        } = arraySchema(property);

        output.properties[propertyKey] = restOfArraySchema;
        break;
      default:
        throw new Error(`${property.type} not implemented`);
    }
  }

  return output;
}

export function stringSchema(parameter: any): SchemaTypeString {
  const supportedFormats = Object.values(StringFormats);
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeString = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.format &&
      supportedFormats.includes(parameter.format) && {
        format: parameter.format
      }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(parameter.minLength && { minLength: parameter.minLength }),
    ...(parameter.maxLength && { maxLength: parameter.maxLength }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "string"
  };

  if (parameter.meta) {
    if (parameter.meta.format) {
      parameter.format = parameter.meta.format;
    }
  }

  if (parameter.format && supportedFormats.includes(parameter.format)) {
    output.format = parameter.format;

    switch (parameter.format) {
      case StringFormats.Date:
        output.maxLength = 10; // 2020-10-14
        output.minLength = 10;
        break;
      case StringFormats.DateTime:
        output.minLength = 16; // 20201014T214403Z
        output.maxLength = 25; // 2020-10-14T21:44:03+00:00
        break;
      case StringFormats.Password:
        delete output.default; // no defaults for passwords
        break;
    }
  }

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}

export function numberSchema(parameter: any): SchemaTypeNumber {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeNumber = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: NumberFormats.Double,
    type: "number"
  };

  return output;
}

export function booleanSchema(parameter: any): SchemaTypeBoolean {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeBoolean = {
    description,
    ...(typeof parameter.default === "boolean" && {
      default: parameter.default
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "boolean"
  };

  return output;
}

export function integerSchema(parameter: any): SchemaTypeInteger {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeInteger = {
    description,
    ...(typeof parameter.default === "number" && {
      default: parameter.default
    }),
    ...(typeof parameter.example === "number" && {
      example: parameter.example
    }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: IntegerFormats.Int64,
    type: "integer"
  };

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}

export function arraySchema(parameter: any): SchemaTypeArray {
  const output: SchemaTypeArray = {
    ...(typeof parameter.default === "object" && {
      default: parameter.default
    }),
    ...(parameter.minItems && { minItems: parameter.minItems }),
    ...(parameter.maxItems && { maxItems: parameter.maxItems }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "array"
  };

  switch (parameter.items.type) {
    case "string":
      const {
        default: ignoreA1,
        nullable: ignoreA2,
        minLength,
        maxLength,
        ...restOfStringSchema
      } = stringSchema(parameter.items);

      output.items = restOfStringSchema;
      break;
    case "number":
      const {
        default: ignoreB1,
        nullable: ignoreB2,
        minimum: ignoreB3,
        maximum: ignoreB4,
        ...restNumberOfSchema
      } = numberSchema(parameter.items);

      output.items = restNumberOfSchema;
      break;
    case "integer":
      const {
        default: ignoreC1,
        nullable: ignoreC2,
        minimum: ignoreC3,
        maximum: ignoreC4,
        ...restIntegerOfSchema
      } = integerSchema(parameter.items);

      output.items = restIntegerOfSchema;
      break;
    case "object":
      const {
        default: ignoreD1,
        nullable: ignoreD2,
        ...restOfObjectSchema
      } = objectSchema(parameter.items);

      output.items = restOfObjectSchema;
      break;
    default:
      throw new Error("not implemented");
  }

  if (parameter.items.enum instanceof Array) {
    // default values must be part of enum
    if (
      output.default instanceof Array &&
      !output.default.every((element: any, _index: number, _array: any[]) =>
        parameter.items.enum.includes(element)
      )
    ) {
      delete output.default;
    }

    // string enums are already limited
    if (parameter.items.type === "string") {
      delete (output.items as SchemaTypeString).minLength;
      delete (output.items as SchemaTypeString).maxLength;
    }
  }

  return output;
}

export function bodySchema(schema: ObjectSchema): Body {
  return bodyParams(schema);
}
