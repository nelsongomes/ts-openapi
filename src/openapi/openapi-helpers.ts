export function textPlain(description: string) {
  return {
    content: {
      "text/plain": {
        schema: {
          type: "string",
        },
      },
    },
    description,
  };
}
