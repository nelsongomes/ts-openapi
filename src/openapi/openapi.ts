import { Schema } from "joi";
import { ApplicationError } from "../errors/application-error";
import joiToSwagger, { SwaggerSchema } from "../joi-conversion";
import {
  bodyParams,
  limitations,
  stringSchema,
  numberSchema,
  integerSchema,
  booleanSchema,
  arraySchema,
} from "./openapi-functions";
import { validateParameters } from "./openapi-validation";
import {
  OpenApiSchema,
  Parameter,
  ParameterIn,
  Parameters,
  Path,
  Servers,
  SecurityScheme,
  WebRequestSchema,
  PathDefinition,
} from "./openapi.types";

const REQUIRED_TYPES = [ParameterIn.Path];

export class OpenApi {
  private schema: OpenApiSchema;
  private operationIds: string[] = [];
  private securitySchemeIds: string[] = [];

  constructor(
    version: string,
    title: string,
    description: string,
    email: string
  ) {
    this.schema = {
      info: {
        contact: {
          email,
        },
        description,
        license: {
          name: "Apache 2.0",
          url: "http://www.apache.org/licenses/LICENSE-2.0.html",
        },
        termsOfService: "http://swagger.io/terms/",
        title,
        version,
      },
      openapi: "3.0.1",
      paths: {},
      servers: [],
    };
  }

  public setServers(servers: Servers) {
    this.schema.servers = servers;
  }

  public declareSecurityScheme(name: string, scheme: SecurityScheme) {
    if (this.securitySchemeIds.includes(name)) {
      throw new Error("Security scheme name already exists.");
    }

    if (!this.schema.components || !this.schema.components.securitySchemes) {
      this.schema.components = {};
      this.schema.components.securitySchemes = {};
    }

    // add new securityScheme by name
    this.schema.components.securitySchemes[name] = scheme;

    // add new id
    this.securitySchemeIds.push(name);
  }

  public addGlobalSecurityScheme(name: string, scopes?: string[]) {
    if (!this.securitySchemeIds.includes(name)) {
      throw new Error(`Unknown security scheme '${name}'`);
    }

    if (!this.schema.security) {
      this.schema.security = [];
    }

    this.schema.security.push({ [name]: scopes || [] });
  }

  public addPath(path: string, definition: Path, visible: boolean) {
    if (!visible) {
      return;
    }

    if (definition.get && definition.get.requestBody) {
      throw new Error("GET operations cannot have a requestBody.");
    }

    const pathDefinition =
      (definition.get as PathDefinition) ||
      (definition.post as PathDefinition) ||
      (definition.put as PathDefinition) ||
      (definition.delete as PathDefinition);

    const operationId = pathDefinition.operationId;
    const responses = pathDefinition.responses;

    if (Object.getOwnPropertyNames(responses).length === 0) {
      throw new Error("Should define at least one response.");
    }

    if (pathDefinition.security) {
      // TODO verify security scheme exists
      pathDefinition.security.forEach((securityScheme) => {
        const securitySchemeName = Object.getOwnPropertyNames(
          securityScheme
        )[0];

        if (!this.securitySchemeIds.includes(securitySchemeName)) {
          throw new Error(`Unknown security scheme '${securitySchemeName}'`);
        }
      });
    }

    if (!operationId) {
      throw new Error("No operationId supplied.");
    }

    if (this.operationIds.includes(operationId)) {
      throw new Error("Operations must have unique operationIds.");
    }

    this.operationIds.push(operationId);
    this.schema.paths[path] = definition;
  }

  public setLicense(name: string, url: string, termsOfService: string) {
    this.schema.info.license.name = name;
    this.schema.info.license.url = url;
    this.schema.info.termsOfService = termsOfService;
  }

  public parametersAndBodyFromSchema(validationSchema?: WebRequestSchema) {
    let parameters: Parameters = [];
    let requestBody;

    if (validationSchema) {
      if (validationSchema.query) {
        // get parameters
        this.genericParams(
          parameters,
          validationSchema.query,
          ParameterIn.Query
        );
      }

      if (validationSchema.params) {
        // uri params
        this.genericParams(
          parameters,
          validationSchema.params,
          ParameterIn.Path
        );
      }

      if (validationSchema.headers) {
        // header params
        this.genericParams(
          parameters,
          validationSchema.headers,
          ParameterIn.Header
        );
      }

      if (validationSchema.body) {
        // request body
        requestBody = bodyParams(validationSchema.body);
      }

      return {
        parameters: parameters.length > 0 ? parameters : undefined,
        requestBody,
      };
    }

    return { parameters: undefined, requestBody: undefined };
  }

  public genericParams(
    parameters: Parameters,
    schema: Schema,
    type: ParameterIn
  ) {
    if (type === ParameterIn.Body) {
      throw new Error("Body content must declared with function bodyParams.");
    }

    const query = joiToSwagger(schema, {});

    for (const key of Object.keys(query.swagger.properties)) {
      const parameter = query.swagger.properties[key];
      const isParameterRequired = this.isRequiredProperty(query.swagger, key);

      // specific parameter context validation
      validateParameters(
        type,
        parameter,
        key,
        parameter.type,
        isParameterRequired
      );

      switch (parameter.type) {
        case "string":
          parameters.push(
            this.stringParameter(key, parameter, isParameterRequired, type)
          );
          break;
        case "number":
        case "integer":
          if (parameter.format && parameter.format === "float") {
            parameters.push(
              this.numberParameter(key, parameter, isParameterRequired, type)
            );
          } else {
            parameters.push(
              this.integerParameter(key, parameter, isParameterRequired, type)
            );
          }
          break;
        case "boolean":
          parameters.push(
            this.booleanParameter(key, parameter, isParameterRequired, type)
          );
          break;
        case "array":
          parameters.push(
            this.arrayParameter(key, parameter, isParameterRequired, type)
          );
          break;
        default:
          throw new Error(`Unknown type '${parameter.type}' in openapi`);
      }
    }
  }

  private isRequiredProperty(swagger: SwaggerSchema, key: string) {
    return swagger.required && swagger.required.includes(key);
  }

  public generateJson(): {} {
    if (this.schema.servers.length === 0) {
      throw new ApplicationError(
        500,
        "No servers were added to OpenApi definition"
      );
    }

    if (Object.getOwnPropertyNames(this.schema.paths).length === 0) {
      throw new ApplicationError(
        500,
        "No paths were added to OpenApi definition."
      );
    }

    return this.schema;
  }

  private stringParameter(
    name: string,
    parameter: any,
    required: boolean,
    type: ParameterIn
  ): Parameter {
    const limitationDetail = limitations(parameter);
    const p: Parameter = {
      description:
        (parameter.description || "Parameter without description.") +
        (limitationDetail || ""),
      in: type,
      name,
      required: isRequiredParameter(required, type),
      schema: stringSchema(parameter),
      ...(parameter.example && { example: parameter.example }),
    };

    return p;
  }

  private numberParameter(
    name: string,
    parameter: any,
    required: boolean,
    type: ParameterIn
  ): Parameter {
    const limitationDetail = limitations(parameter);

    const p: Parameter = {
      description:
        (parameter.description || "Parameter without description.") +
        (limitationDetail || ""),
      in: type,
      name,
      required: isRequiredParameter(required, type),
      schema: numberSchema(parameter),
      ...(parameter.example && { example: parameter.example }),
    };

    return p;
  }

  private integerParameter(
    name: string,
    parameter: any,
    required: boolean,
    type: ParameterIn
  ): Parameter {
    const limitationDetail = limitations(parameter);

    const p: Parameter = {
      description:
        (parameter.description || "Parameter without description.") +
        (limitationDetail || ""),
      in: type,
      name,
      required: isRequiredParameter(required, type),
      schema: integerSchema(parameter),
      ...(parameter.example && { example: parameter.example }),
    };

    return p;
  }

  private booleanParameter(
    name: string,
    parameter: any,
    required: boolean,
    type: ParameterIn
  ): Parameter {
    const limitationDetail = limitations(parameter);

    const p: Parameter = {
      description:
        (parameter.description || "Parameter without description.") +
        (limitationDetail || ""),
      in: type,
      name,
      required: isRequiredParameter(required, type),
      schema: booleanSchema(parameter),
      ...(typeof parameter.example === "boolean" && {
        example: parameter.example,
      }),
    };

    return p;
  }

  private arrayParameter(
    name: string,
    parameter: any,
    required: boolean,
    type: ParameterIn
  ): Parameter {
    const limitationDetail = limitations(parameter);
    const schema = arraySchema(parameter);

    const p: Parameter = {
      description:
        (parameter.description || "Parameter without description.") +
        (limitationDetail || ""),
      in: type,
      name,
      required: isRequiredParameter(required, type),
      schema,
      ...(parameter.example && { example: parameter.example }),
    };

    return p;
  }
}

// all required parameters or path parameters are parameters are required
function isRequiredParameter(required: boolean, type: ParameterIn) {
  return required || REQUIRED_TYPES.includes(type);
}
