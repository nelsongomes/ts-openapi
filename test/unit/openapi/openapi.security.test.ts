import { OpenApi } from "../../../src/openapi/openapi";
import {
  apiKeyAuth,
  basicAuth,
  bearerAuth,
  cookieAuth,
  textPlain,
} from "../../../src/openapi/helpers/openapi-helpers";

describe("src/openapi/openapi", () => {
  describe("Security Schemes", () => {
    test("Basic Security (global)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme("basicSecurity", basicAuth());

        // declare global schemes (applicable to all methods)
        openApi.addGlobalSecurityScheme("basicSecurity");

        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              parameters: [],
              responses: {
                200: textPlain("Successful operation."),
              },
              summary: "Server Healthcheck",
              tags: ["Internals"],
            },
          },
          true
        );

        expect(openApi.generateJson()).toMatchSnapshot();
      } catch (e) {
        fail("No exception expected");
      }
    });

    test("Basic Security (local)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme("basicSecurity", basicAuth());

        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              parameters: [],
              responses: {
                200: textPlain("Successful operation."),
              },
              summary: "Server Healthcheck",
              tags: ["Internals"],
              security: [{ basicSecurity: [] }],
            },
          },
          true
        );

        expect(openApi.generateJson()).toMatchSnapshot();
      } catch (e) {
        fail("No exception expected");
      }
    });

    test("Api Key Security (local)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme(
          "apiSecurity",
          apiKeyAuth("X-API-KEY", "header")
        );

        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              parameters: [],
              responses: {
                200: textPlain("Successful operation."),
              },
              summary: "Server Healthcheck",
              tags: ["Internals"],
              security: [{ apiSecurity: [] }],
            },
          },
          true
        );

        expect(openApi.generateJson()).toMatchSnapshot();
      } catch (e) {
        fail("No exception expected");
      }
    });

    test("Cookie Security (local)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme(
          "cookieSecurity",
          cookieAuth("JSESSIONID")
        );

        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              parameters: [],
              responses: {
                200: textPlain("Successful operation."),
              },
              summary: "Server Healthcheck",
              tags: ["Internals"],
              security: [{ cookieSecurity: [] }],
            },
          },
          true
        );

        expect(openApi.generateJson()).toMatchSnapshot();
      } catch (e) {
        fail("No exception expected");
      }
    });

    test("Bearer Security (local)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme("bearerSecurity", bearerAuth());

        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              parameters: [],
              responses: {
                200: textPlain("Successful operation."),
              },
              summary: "Server Healthcheck",
              tags: ["Internals"],
              security: [{ bearerSecurity: [] }],
            },
          },
          true
        );

        expect(openApi.generateJson()).toMatchSnapshot();
      } catch (e) {
        fail("No exception expected");
      }
    });

    test("Bearer Security with bearerFormat (local)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme("bearerJwtSecurity", bearerAuth("JWT"));

        openApi.addPath(
          "/health",
          {
            get: {
              description: "Service healthcheck endpoint",
              operationId: "healthcheck",
              parameters: [],
              responses: {
                200: textPlain("Successful operation."),
              },
              summary: "Server Healthcheck",
              tags: ["Internals"],
              security: [{ bearerJwtSecurity: [] }],
            },
          },
          true
        );

        expect(openApi.generateJson()).toMatchSnapshot();
      } catch (e) {
        fail("No exception expected");
      }
    });

    test("Should throw an error (global)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme("basicSecurity", basicAuth());

        // declare global schemes (applicable to all methods)
        openApi.addGlobalSecurityScheme("basicSecurityUnknown");
      } catch (e) {
        expect(e.message).toBe(
          "Unknown security scheme 'basicSecurityUnknown'"
        );
      }
    });

    test("Should throw an error (local)", async () => {
      const openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      try {
        openApi.setServers([{ url: "https://server.com" }]);

        // declare security schemes available
        openApi.declareSecurityScheme("basicSecurity", basicAuth());
      } catch (e) {
        expect(e.message).toBe(
          "Unknown security scheme 'basicSecurityUnknown'"
        );
      }
    });
  });
});
