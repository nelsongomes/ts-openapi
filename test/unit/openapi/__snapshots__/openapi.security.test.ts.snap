// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/openapi/openapi Security Schemes Api Key Security (local) 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "apiSecurity": Object {
        "in": "header",
        "name": "X-API-KEY",
        "type": "apiKey",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "healthcheck",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "security": Array [
          Object {
            "apiSecurity": Array [],
          },
        ],
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Security Schemes Basic Security (global) 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "basicSecurity": Object {
        "scheme": "basic",
        "type": "http",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "healthcheck",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "security": Array [
    Object {
      "basicSecurity": Array [],
    },
  ],
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Security Schemes Basic Security (local) 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "basicSecurity": Object {
        "scheme": "basic",
        "type": "http",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "healthcheck",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "security": Array [
          Object {
            "basicSecurity": Array [],
          },
        ],
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Security Schemes Bearer Security (local) 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "bearerSecurity": Object {
        "scheme": "bearer",
        "type": "http",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "healthcheck",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "security": Array [
          Object {
            "bearerSecurity": Array [],
          },
        ],
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Security Schemes Bearer Security with bearerFormat (local) 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "bearerJwtSecurity": Object {
        "bearerFormat": "JWT",
        "scheme": "bearer",
        "type": "http",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "healthcheck",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "security": Array [
          Object {
            "bearerJwtSecurity": Array [],
          },
        ],
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Security Schemes Cookie Security (local) 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "cookieSecurity": Object {
        "in": "cookie",
        "name": "JSESSIONID",
        "type": "apiKey",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "healthcheck",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text-plain": Object {},
            },
            "description": "Successful operation.",
          },
        },
        "security": Array [
          Object {
            "cookieSecurity": Array [],
          },
        ],
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;
